idea/src/org/jetbrains/kotlin/idea/joinLines/JoinToStringTemplateHandler.kt

```diff
@@ -35,16 +35,66 @@ class JoinToStringTemplateHandler : JoinRawLinesHandlerDelegate {
         val index = if (c == '\n') start - 1 else start
 
         val plus = file.findElementAt(index)?.takeIf { it.node?.elementType == KtTokens.PLUS } ?: return -1
-        val expression = (plus.parent.parent as? KtBinaryExpression) ?: return -1
-        val left = expression.left ?: return -1
-        val right = expression.right ?: return -1
-        if (left !is KtStringTemplateExpression || right !is KtStringTemplateExpression) return -1
-        if (!ConvertToStringTemplateIntention.isApplicableToNoParentCheck(expression)) return -1
-
-        val offset = left.endOffset - 1
-        expression.replace(ConvertToStringTemplateIntention.buildReplacement(expression))
-        return offset
+        val binaryExpr = (plus.parent.parent as? KtBinaryExpression) ?: return -1
+        if (!binaryExpr.joinable()) return -1
+
+        val lineCount = binaryExpr.getLineCount()
+        val targetBinaryExpr = binaryExpr.expand(lineCount + 1)
+        val (left, rightText) = targetBinaryExpr.collect(lineCount)
```

Comment:
In general, it seems to work fine. But can you come up with the better name for `expand` and `collect` functions? Also I would suggest rewriting `expand` and `collect` without recursion.

Reply:
Fixed